sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerQueryTestSuite.scala

```diff
@@ -0,0 +1,359 @@
+            }
+        }
+      }
+    }
+  }
+
+  override def createScalaTestCase(testCase: TestCase): Unit = {
+    if (blackList.exists(t =>
+      testCase.name.toLowerCase(Locale.ROOT).contains(t.toLowerCase(Locale.ROOT)))) {
+      // Create a test case to ignore this case.
+      ignore(testCase.name) { /* Do nothing */ }
+    } else {
+      // Create a test case to run this case.
+      test(testCase.name) {
+        runTest(testCase)
+      }
+    }
+  }
+
+  override def listTestCases(): Seq[TestCase] = {
+    listFilesRecursively(new File(inputFilePath)).flatMap { file =>
+      val resultFile = file.getAbsolutePath.replace(inputFilePath, goldenFilePath) + ".out"
+      val absPath = file.getAbsolutePath
+      val testCaseName = absPath.stripPrefix(inputFilePath).stripPrefix(File.separator)
+
+      if (file.getAbsolutePath.startsWith(s"$inputFilePath${File.separator}udf")) {
+        Seq.empty
+      } else if (file.getAbsolutePath.startsWith(s"$inputFilePath${File.separator}pgSQL")) {
+        PgSQLTestCase(testCaseName, absPath, resultFile) :: Nil
+      } else {
+        RegularTestCase(testCaseName, absPath, resultFile) :: Nil
+      }
+    }
+  }
+
+  test("Check if ThriftServer can work") {
+    withJdbcStatement { statement =>
+      val rs = statement.executeQuery("select 1L")
+      rs.next()
+      assert(rs.getLong(1) === 1L)
+    }
+  }
+
+  private def getNormalizedResult(statement: Statement, sql: String): Seq[String] = {
+    try {
+      val rs = statement.executeQuery(sql)
+      val cols = rs.getMetaData.getColumnCount
+      val buildStr = () => (for (i <- 1 to cols) yield {
+        getHiveResult(rs.getObject(i))
+      }).mkString("\t")
+
+      val answer = Iterator.continually(rs.next()).takeWhile(identity).map(_ => buildStr()).toSeq
+        .map(replaceNotIncludedMsg)
+      if (isNeedSort(sql)) {
+        answer.sorted
+      } else {
+        answer
+      }
+    } catch {
+      case a: AnalysisException =>
+        // Do not output the logical plan tree which contains expression IDs.
+        // Also implement a crude way of masking expression IDs in the error message
+        // with a generic pattern "###".
+        val msg = if (a.plan.nonEmpty) a.getSimpleMessage else a.getMessage
+        Seq(a.getClass.getName, msg.replaceAll("#\\d+", "#x")).sorted
+      case NonFatal(e) =>
+        // If there is an exception, put the exception class followed by the message.
+        Seq(e.getClass.getName, e.getMessage)
+    }
+  }
+
+  private def startThriftServer(port: Int, attempt: Int): Unit = {
+    logInfo(s"Trying to start HiveThriftServer2: port=$port, attempt=$attempt")
+    val localSparkSession = spark.newSession()
+    val sqlContext = localSparkSession.sqlContext
+    sqlContext.setConf(ConfVars.HIVE_SERVER2_THRIFT_PORT.varname, port.toString)
+    hiveServer2 = HiveThriftServer2.startWithContext(sqlContext)
+  }
+
+  private def withJdbcStatement(fs: (Statement => Unit)*) {
```

Comment:
Could we keep the connection open, to not have to start a sessopm amd reload the test data into temp views each time?
We could open the connection with `conn = DriverManager.getConnection(jdbcUri, user, "")` in beforeAll, load the test data there, and then have `withJDBCStatement` just create new statements, finally closing the connection in afterAll.

However, it seems that opening new connection/session may be by design here, for test isolation. Then we'll have to leave it as is, but I think we should still be able to avoid starting the ThriftServer beforeEach.

Reply:
If move `loadTestData` to `beforeAll`. Some tests will fail:
![image](https://user-images.githubusercontent.com/5399861/62935611-2747c780-bdfa-11e9-8974-4ac79696783b.png)